src/mlpack/core/data/dictionary_encoding_impl.hpp

```diff
@@ -0,0 +1,106 @@
+/**
+ * @file dictionary_encoding_impl.hpp
+ * @author Jeffin Sam
+ *
+ * Implementation of dictionary encoding functions.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_DICT_ENCODING_IMPL_HPP
+#define MLPACK_CORE_DATA_DICT_ENCODING_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "dictionary_encoding.hpp"
+
+namespace mlpack {
+namespace data {
+
+void DicitonaryEncoding::Reset()
+{
+  mappings.clear();
+}
+
+template<typename TokenizerType>
+void DicitonaryEncoding::CreateMap(std::string& strings,
+    TokenizerType tokenizer)
+{
+  boost::string_view strView(strings);
+  boost::string_view token;
+  token = tokenizer(strView);
+  std::size_t curLabels = mappings.size() + 1;
+  while (!token.empty())
+  {
+    if (mappings.find(token) == mappings.end())
+    {
+      mappings[std::move(token)] = curLabels++;
+    }
+    token = tokenizer(strView);
+  }
+}
+
+template<typename MatType, typename TokenizerType>
+void DicitonaryEncoding::Encode(const std::vector<std::string>& strings,
+                MatType& output, TokenizerType tokenizer)
+{
+  boost::string_view strView;
+  boost::string_view token;
+  std::vector< std::vector<boost::string_view> > dataset;
+  size_t colSize = 0;
+  for (size_t i = 0; i < strings.size(); i++)
+  {
+    strView = strings[i];
+    token = tokenizer(strView);
+    dataset.push_back(std::vector<boost::string_view>() );
+    while (!token.empty())
+    {
+      dataset[i].push_back(token);
+      token = tokenizer(strView);
+    }
+    colSize = std::max(colSize, dataset[i].size());
+  }
+  size_t curLabel = mappings.size() + 1;
+  output.zeros(dataset.size(), colSize);
+  for (size_t i = 0; i < dataset.size(); ++i)
+  {
+    for (size_t j = 0; j < dataset[i].size(); ++j)
+    {
+      if (mappings.count(dataset[i][j]) == 0)
+      {
+        mappings[dataset[i][j]] = curLabel++;
```

Comment:
Looks like there is an error. `dataset[i][j]` is a `string_view` that refers to the `strings` argument. But the life time of the argument could be less than the life time of `mappings` since `mappings` is a class field. If you passed a temporary vector to `Encode()`, you wouldn't be able to use the map after the function call.

I suggest to copy each token which you want to insert to the map. For example you could copy all tokens to a vector (the vector must be a class field). And then you should insert the `string_view` of the copied string to the map.

Reply:
That was very good point. Made changes.